src/rgw/rgw_data_sync.cc

```diff
@@ -4136,12 +4138,78 @@ static bool has_olh_epoch(RGWModifyOp op) {
   return op == CLS_RGW_OP_LINK_OLH || op == CLS_RGW_OP_UNLINK_INSTANCE;
 }
 
+class RGWBucketShardIsDoneCR: public RGWCoroutine {
+  RGWDataSyncCtx *sc;
+  RGWDataSyncEnv *sync_env;
+  rgw_bucket_sync_status bucket_status;
+  const rgw_raw_obj status_obj;
+  const uint64_t num_shards;
+  const int shard_id;
+  RGWObjVersionTracker objv_tracker;
+  std::optional<next_bilog_result>& next_log;
```

Comment:
can this reference be const?

Reply:
err, is this `next_bilog_result` actually optional here? we shouldn't be calling `RGWBucketShardIsDoneCR` unless we know there's a new generation to switch to

Reply:
yes, we call RGWBucketShardIsDoneCR only if next generation exists.. fixed.